counsel.el

```diff
@@ -6317,20 +6318,25 @@ list is passed to `compilation-environment'."
 (defvar counsel-compile-phony-pattern "^\\.PHONY:[\t ]+\\(.+\\)$"
   "Regexp for extracting phony targets from Makefiles.")
 
+(defvar counsel-compile-help-pattern
+  "\\(?:^\\(\\*?\\)[[:space:]]+\\([[:alnum:]]+\\)[[:space:]]+-\\)"
+  "Regexp for extracting help targets from a make help call.")
+
 ;; This is loosely based on the Bash Make completion code
 (defun counsel-compile--probe-make-targets (dir)
   "Return a list of Make targets for DIR.
 
-Return an empty list is Make exits with an error.  This might
-happen because some sort of configuration needs to be done first
-or the source tree is pristine and being used for multiple build
-trees."
+Return an single blank target (so we invoke the default target)
+if Make exits with an error.  This might happen because some sort
+of configuration needs to be done first or the source tree is
+pristine and being used for multiple build trees."
   (let ((default-directory dir)
         (targets nil))
     (with-temp-buffer
       ;; 0 = no-rebuild, -q & 1 needs rebuild, 2 error (for GNUMake at
       ;; least)
-      (when (< (call-process "make" nil t nil "-nqp") 2)
+      (if (> (call-process "make" nil t nil "-nqp") 1)
```

Comment:
`call-process` does not always return a number.

Reply:
I take it this would have failed before? What would you recommend - a cond form with multiple legs handling the various cases?

Reply:
How about:

      (let ((res (call-process "make" nil t nil "-nqp")))
        (cond ((and (numberp res)
                    (< res 2))
               (goto-char (point-min))
               (while (re-search-forward counsel-compile-phony-pattern nil t)
                 (setq targets
                       (nconc targets (split-string
                                       (match-string-no-properties
                                        1))))))
              (t (setq targets '(""))))))

Reply:
> I take it this would have failed before?

Yep.

> What would you recommend - a cond form with multiple legs handling the various cases?

I would recommend `cond` whenever there is more than one condition or fall-through case.  If there is only one condition you can just use `if`.

> How about:

Passing a constant literal such as `'("")` to a destructive function such as `sort` is undefined behaviour.  Instead, you can write something like:

```el
(with-temp-buffer
  (let* ((default-directory dir)
         (res (call-process "make" nil t nil "-nqp")))
    (if (and (wholenump res) (< res 2))
        (let (targets)
          (goto-char (point-min))
          (while (re-search-forward counsel-compile-phony-pattern nil t)
            (push (split-string (match-string-no-properties 1)) targets))
          (sort (apply #'nconc targets) #'string-lessp))
      (list ""))))
```

or:

```el
(with-temp-buffer
  (let* ((default-directory dir)
         (res (call-process "make" nil t nil "-nqp"))
         targets)
    (if (or (not (numberp res)) (> res 1))
        (list "")
      (goto-char (point-min))
      (while (re-search-forward counsel-compile-phony-pattern nil t)
        (push (split-string (match-string-no-properties 1)) targets))
      (sort (apply #'nconc targets) #'string-lessp))))
```